src/test/java/com/karankumar/bookproject/ui/book/BookFormTest.java

```diff
@@ -97,38 +103,49 @@ public void setup() {
         customShelfService.save(customShelf);
 
         Assumptions.assumeTrue(predefinedShelfService != null);
-        bookForm = createBookForm(true);
-
+        bookForm = createBookForm(READ, true);
     }
 
-    private BookForm createBookForm(boolean isInSeries) {
+    private BookForm createBookForm(ShelfName shelf, boolean isInSeries) {
         BookForm bookForm = new BookForm(predefinedShelfService, customShelfService);
-        bookForm.setBook(createBook(predefinedShelfService, isInSeries));
+        readShelf = predefinedShelfService.findAll().get(2);
+        bookForm.setBook(createBook(shelf, isInSeries));
         return bookForm;
-
     }
 
-    private Book createBook(PredefinedShelfService predefinedShelfService, boolean isInSeries) {
+    private Book createBook(ShelfName shelfName, boolean isInSeries) {
+        // values that are present in all forms
         Author author = new Author(firstName, lastName);
-        readShelf = predefinedShelfService.findAll().get(2);
-        Book book = new Book(bookTitle, author, readShelf);
-
+        PredefinedShelf shelf = predefinedShelfService.findAll(shelfName).get(0);
+        Book book = new Book(bookTitle, author, shelf);
 
         pagesRead = generateRandomNumberOfPages();
         numberOfPages = generateRandomNumberOfPages();
         seriesPosition = SERIES_POSITION;
 
-        book.setCustomShelf(customShelf);
         book.setGenre(genre);
-        book.setPagesRead(pagesRead);
         book.setNumberOfPages(numberOfPages);
-        book.setDateStartedReading(dateStarted);
-        book.setDateFinishedReading(dateFinished);
-        book.setRating(ratingVal);
+        book.setCustomShelf(customShelfService.findAll().get(0));
         if (isInSeries) {
-            book.setSeriesPosition(seriesPosition);
+            book.setSeriesPosition(SERIES_POSITION);
         }
 
+        // values that are only present for specific predefined shelves
+        switch (shelfName) {
+            case READING:
+                book.setDateStartedReading(dateStarted);
+                break;
+            case READ:
+                book.setDateStartedReading(dateStarted);
+                book.setDateFinishedReading(dateFinished);
+                book.setRating(ratingVal);
+                break;
+            case DID_NOT_FINISH:
+                book.setDateStartedReading(dateStarted);
+                book.setPagesRead(pagesRead);
+                book.setRating(ratingVal);
+                break;
+        }
```

Comment:
This works and is OK, but I think it would be nice to refactor this out to a method (called, e.g., `setOptionalBookFieldsBasedOnShelf` (a shorter but equally or more descriptive name would be nice)). Then, we can do away with the comment.

I'm thinking that this method can take a `Book` as an argument and return a `Book` (that has the optional fields set). You could then return this. 

What do you think?

Reply:
Sounds good to me. :)
Would you agree with this name?
```java
private Book setShelfSpecificFields(Book book, PredefinedShelf.ShelfName shelfName)
```

Reply:
That's good

Reply:
done